addons/l10n_hu_edi/models/l10n_hu_edi_transaction.py

```diff
@@ -0,0 +1,531 @@
+                    fn.__name__,
+                    set(self.mapped('state')),
+                    start_states,
+                ))
+            fn(self, *args, **kwargs)
+            if any(t.state not in end_states for t in self):
+                _logger.error(_(
+                    'Recordset %s: Method %s returned with bad end state(s): %s, allowed states are %s',
+                    self,
+                    fn.__name__,
+                    set(self.mapped('state')),
+                    end_states,
+                ))
+        return decorated_fn
+    return decorate
+
+
+class L10nHuEdiTransaction(models.Model):
+    #################################################################################################
+    # OVERVIEW
+    #
+    # * This model works as a state machine.
+    # * A transaction represents an invoice object on NAV's side.
+    # * New transactions are created in the 'to_send' state.
+    # * The state can be advanced through the `upload`, `abort` and `query_status` methods.
+    # * Some states are 'active': in these states, the invoice number is reserved on NAV's side,
+    #       therefore a new transaction must not be created for the invoice.
+    # * Some states are final, which means that once the transaction has arrived into them, they can
+    #       no longer be updated.
+    # * Non-active final states are not blocking: the user can reset the invoice to draft,
+    # * Active final states are blocking: in those states, the user can no longer delete the invoice,
+    #       and must create a credit note it if he wants to annul it.
+    #
+    # STATE DIAGRAM
+    # * to_send, token_error --[upload]--> token_error, sent, send_error, send_timeout
+    # * to_send, token_error --[abort]--> unsent
+    # * sent, query_error --[query_status]--> sent, confirmed, confirmed_warning, rejected, query_error
+    # * send_timeout --[recover_timeout]--> to_send, sent, send_timeout
+    # * send_error: is a final state due to recording requirements (to retry, create a new transaction)
+    # * confirmed, confirmed_warning, rejected, unsent: are final states
+    #################################################################################################
+    _name = 'l10n_hu_edi.transaction'
+    _description = 'Hungarian Tax Authority Invoice Upload Transaction'
+    _order = 'create_date desc'
+    _rec_name = 'id'
+
+    state = fields.Selection(
+        selection=[
+            ('unsent', 'Unsent (aborted)'),
+            ('to_send', 'To Send'),
+            ('token_error', 'To Retry, could not get an authentication token'),
+            ('sent', 'Sent, waiting for response'),
+            ('send_error', 'Error when sending'),
+            ('send_timeout', 'Timeout when sending'),
+            ('confirmed', 'Confirmed'),
+            ('confirmed_warning', 'Confirmed with warnings'),
+            ('rejected', 'Rejected'),
+            ('query_error', 'To Retry, error when requesting status'),
```

Comment:
This state diagram with active/non-active states, can_perform, ... is overkill.  It seems to be done mainly for the timeout, which probably if you send an invoice just a second time, you could find out there had been a timeout before.  However, it seems that it is important that if we have a "rejected" state that we can not reset to draft.